sw/device/tests/dif/dif_padctrl_test.cc

```diff
@@ -0,0 +1,756 @@
+              kDifPadctrlChangeBadArg);
+  }
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrsFailureLocked) {
+  DeviceInit();
+
+  ExpectLockCheck(true);
+
+  EXPECT_EQ(dif_padctrl_write_attrs(&dif_padctrl_, kDifPadctrlPadMio, 0,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlChangeLocked);
+}
+
+TEST_F(DifPadctrlTest, EnableAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_enable_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                     kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, DisableAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_disable_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                      kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, WriteAttrsFailureInvalidPadctrl) {
+  DeviceInit();
+
+  EXPECT_EQ(dif_padctrl_write_attrs(nullptr, kDifPadctrlPadMio, 0,
+                                    kDifPadctrlAttrPullUp),
+            kDifPadctrlBadArg);
+}
+
+TEST_F(DifPadctrlTest, GetAttrsSuccess) {
+  DeviceInit();
+
+  for (auto attr_field : mio_attr_fields_) {
+    dif_padctrl_attr_t expected_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrPullDown | kDifPadctrlAttrIOInvert);
+    ExpectAttrFieldRead(attr_field, expected_attrs);
+
+    dif_padctrl_attr_t seen_attrs = kDifPadctrlAttrNone;
+    EXPECT_EQ(dif_padctrl_get_attrs(&dif_padctrl_, kDifPadctrlPadMio,
+                                    attr_field.pad_num, &seen_attrs),
+              kDifPadctrlOk);
+
+    EXPECT_EQ(expected_attrs, seen_attrs);
+  }
+
+  for (auto attr_field : dio_attr_fields_) {
+    dif_padctrl_attr_t expected_attrs =
+        dif_padctrl_attr_t(kDifPadctrlAttrWeakDrive | kDifPadctrlAttrKeeper |
+                           kDifPadctrlAttrIOInvert);
+    ExpectAttrFieldRead(attr_field, expected_attrs);
+
+    dif_padctrl_attr_t seen_attrs = kDifPadctrlAttrNone;
+    EXPECT_EQ(dif_padctrl_get_attrs(&dif_padctrl_, kDifPadctrlPadDio,
+                                    attr_field.pad_num, &seen_attrs),
+              kDifPadctrlOk);
+
+    EXPECT_EQ(expected_attrs, seen_attrs);
+  }
+}
+
+TEST_F(DifPadctrlTest, GetAttrsFailureBadArg) {
+  dif_padctrl_attr_t seen_attrs;
+
+  EXPECT_EQ(dif_padctrl_get_attrs(nullptr, kDifPadctrlPadMio, 0, &seen_attrs),
+            kDifPadctrlBadArg);
+
+  DeviceInit();
```

Comment:
Is there a reason `DeviceInit()` can't be called at the top? I don't think it's actually relevant to the difference between the two `EXPECT_EQ` calls, so moving it up clarifies that to the reader.